compiler/qsc_circuit/src/circuit_to_qsharp.rs

```diff
@@ -0,0 +1,294 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#[cfg(test)]
+mod tests;
+
+use regex_lite::{Captures, Regex};
+use rustc_hash::FxHashMap;
+
+use crate::{
+    circuit::{CircuitGroup, Measurement, Unitary},
+    Circuit, Operation,
+};
+
+pub fn circuits_to_qsharp(file_name: String, circuits_json: String) -> Result<String, String> {
+    match serde_json::from_str::<CircuitGroup>(circuits_json.as_str()) {
+        Ok(circuits) => Ok(build_circuits(file_name, circuits.circuits)),
+        Err(e) => Err(format!("Error: {}", e)),
+    }
+}
+
+pub fn build_circuits(file_name: String, circuits: Vec<Circuit>) -> String {
+    if circuits.len() == 1 {
+        build_operation_def(file_name, &circuits[0])
+    } else {
+        let mut qsharp_str = String::new();
+        for (index, circuit) in circuits.iter().enumerate() {
+            let circuit_name = format!("{file_name}{index}");
+            let circuit_str = build_operation_def(circuit_name, circuit);
+            qsharp_str.push_str(&circuit_str);
+        }
+        qsharp_str
+    }
+}
+
+pub fn build_operation_def(circuit_name: String, circuit: &Circuit) -> String {
```

Comment:
This function got pretty long, it's probably worth refactoring it down into smaller logical parts to make it easier to reason about.

Reply:
refactored to have smaller functions